src/sardana/taurus/qt/qtgui/extra_macroexecutor/macrobutton.py

```diff
@@ -329,97 +329,138 @@ def abort(self):
-
-    # Change macro name
-    Qt.QObject.connect(macro_name, Qt.SIGNAL('textChanged(QString)'), mb.setMacroName)
-    Qt.QObject.connect(macro_name, Qt.SIGNAL('textChanged(QString)'), mb.setButtonText)
+        def toggle_progress(self, toggle):
+            visible = self.show_progress.isChecked()
+            self.mb.toggleProgress(visible or toggle)
     
-    # connect the argument editors
-    signals = [(e, 'textChanged(QString)') for e in _argEditors]
-    mb.connectArgEditors(signals)
-
-    def update_result(result):
-        result_label.setText(str(result))
-
-    def toggle_progress(showProgress):
-        visible = show_progress.isChecked()
-        mb.toggleProgress(visible)
-
-    # Toggle progressbar
-    Qt.QObject.connect(show_progress, Qt.SIGNAL('stateChanged(int)'), toggle_progress)
-    # Update possible macro result
-    Qt.QObject.connect(mb, Qt.SIGNAL('resultUpdated'), update_result)
-
-    # Obtain a demo motor
-    try:
-        demo_motor_name = SarDemoEnv(door_name).getMotors()[0]
-    except Exception, e:
-        from taurus.core.util.log import warning, debug
-        warning('It was unable to obtain a demo motor')
-        debug('Details: %s' % e.message)
-        demo_motor_name = ''
-
-    # Since everything is now connected, the parameters will be updated
-    macro_name.setText('ascan')
-    macro_params = [demo_motor_name, '0', '1', '5', '.1']
-    for e,v in zip(_argEditors, macro_params):
-        e.setText(v)
-
+        def getMacroInfo(self, macro_name):
+    
+            from sardana.taurus.qt.qtcore.tango.sardana.macroserver import registerExtensions
+            registerExtensions()
+            door = taurus.Device(self.door_name)
+            try:
+                pars = door.macro_server.getMacroInfoObj(macro_name).parameters
+            except AttributeError as e:
+                print "Macro %s does not exists!" % macro_name
+                return None
+
+            param_names = []
+            default_values = []
+            for p in pars:
+                param_names.append(p['name'])
+                default_values.append(p['default_value'])
+
+            self.macro_name = macro_name 
+            return param_names, default_values  
+
+        def clean_layout(self, layout):
+            
+            while layout.count():
+                child = layout.takeAt(0)
+                if child.widget():
+                    child.widget().deleteLater()
+
+        def create_layout(self, macro_name):
+            p_names, d_values = self.getMacroInfo(macro_name)
+            # Create TOP layout
+            self.w_arg = Qt.QWidget()
+            self.w_arg.setLayout(Qt.QGridLayout())
+            col = 0
+            self.w_arg.layout().addWidget(Qt.QLabel('macro name'), 0, col)
+            self.w_macro_name = Qt.QLineEdit()
+            self.w_arg.layout().addWidget(self.w_macro_name, 1, col)
+
+            _argEditors = []
+            for name in p_names:
```

Comment:
This does not consider param repeats composed from more than one params.

In general I would limit the purpose of this widget to very simple cases of macros with repeat parameters:
* only one repeat parameter (no nested repeat parameters are allowed)
* repeat parameter located as the last one

In case of repeat parameters with more than one param e.g. `mv` macro the widget should populate two `QLineEdit` instances, one for the moveable and one for the position. Now it populates only one of them.